internal/backend/remote-state/s3/client.go

```diff
@@ -334,27 +387,89 @@ func (c *RemoteClient) Unlock(id string) error {
 	ctx, baselog := baselogging.NewHcLogger(ctx, log)
 	ctx = baselogging.RegisterLogger(ctx, baselog)
 
-	if c.ddbTable == "" {
+	if !c.useLockFile && c.ddbTable == "" {
 		return nil
 	}
 
 	lockErr := &statemgr.LockError{}
 
 	log.Info("Unlocking remote state")
 
+	if c.useLockFile {
+		err := c.unlockWithFile(ctx, id, log)
+		if err != nil {
+			lockErr.Err = err
+			return lockErr
+		}
+
+		return nil
+	} else {
+		err := c.unlockWithDynamoDB(ctx, id, lockErr, log)
+		if err != nil {
+			lockErr.Err = err
+			return lockErr
+		}
+	}
+
+	return nil
+}
+
+// unlockWithFile attempts to unlock the remote state by deleting the lock file from Amazon S3.
+//
+// This method is used when the S3 native locking mechanism is in use, which uses a `.tflock` file
+// to manage state locking. The function deletes the lock file to release the lock, allowing other
+// Terraform clients to acquire the lock on the same state file.
+func (c *RemoteClient) unlockWithFile(ctx context.Context, id string, log hclog.Logger) error {
+	// Retrieve the lock file to get its ETag, which is used to verify the lock ID.
+	getOutput, err := c.s3Client.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(c.bucketName),
+		Key:    aws.String(c.lockFilePath),
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to retrieve lock file: %w", err)
+	}
+
+	// Clean up extra quotes from the ETag value.
+	// ETag values in the GetObject response are always quoted, so we remove the quotes to ensure accurate comparison.
+	eTag := aws.ToString(getOutput.ETag)
+	eTag = strings.Trim(eTag, `"`)
+
+	// Verify that the provided lock ID matches the ETag of the retrieved lock file.
+	// When the lock is acquired, the lock ID is set to the ETag of the lock file. During the unlock operation,
+	// we retrieve the lock file from S3 and compare its ETag to the provided lock ID. This ensures that the lock
+	// is only released if the lock ID matches the ETag of the exact lock file that is intended to be unlocked.
+	if eTag != id {
+		return fmt.Errorf("lock ID %s does not match existing lock (%s)", id, eTag)
+	}
```

Comment:
Should this still attempt to delete the appropriate version of the object, even if the latest object ETag does not match? (Specifically the [`DeleteObject` method](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/s3#DeleteObjectInput) allows a `VersionId` argument).